File: sim.py
#!/usr/bin/env python
import sys
import re
import random
import argparse
parser = argparse.ArgumentParser(description='This program simulates reads from a given genome')
parser.add_argument( '-l', "--read_length", type=int, help='expected read length', required=True)
parser.add_argument( '-n', "--read_number", type=int, help='expected number of readsh', required=True)
parser.add_argument( '-e', "--error_rate", type=float, help='expected error rate', required=True)
parser.add_argument( '-f', "--genome", type=str, help='genome fasta file', required=True)
args = parser.parse_args()
read_length = args.read_length
read_number = args.read_number
error_rate = args.error_rate
genome = args.genome

dna = ""
with open(genome,"rt") as f:
     for line in f:
         if not re.match(">",line):
            line = line.rstrip("\n")
            dna = dna + line
basecount = len(dna)

def variant(base):
    baselist = ["A","T","G","C"]
    if base == "N":
       return base
    else:
        baselist.remove(base)
        sub = (random.choice(baselist))
        return sub
snv = (error_rate * read_length * read_number)
no_of_reads = read_number

def writeFile(of):
    of.write("@HISEQ/1:" + str(i+1) + "\n")
    of.write("".join(fasta) + "\n")
    of.write("+" + "\n")
    of.write("?" * read_length + "\n")

with open("read.fq", "w") as of1:
    with open('readSNV.fq', 'w') as of2:
         with open('pos.bed', 'w') as of3:
            for i in range(no_of_reads):
                pos = (random.randrange(basecount-read_length))
                of3.write("1" + "\t" + str(pos) + "\t" + str(pos+50) + "\t+" + "\n")
                fasta = list(dna[pos:pos+read_length])
                writeFile(of1)
                if i < snv:
                   ranpos = random.randrange(read_length -1)
                   nuc = fasta[ranpos]
                   fasta[ranpos] = variant(nuc)
                   writeFile(of2)
                else:
                    writeFile(of2)
sys.exit(0)

File: sim.pl
#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Long;
use Pod::Usage;
my $genome_fasta;
my $read_length;
my $read_number;
my $error_rate;
my $help;
my $man;
GetOptions('genome_fasta=s'  => \$genome_fasta,
           'read_length=i'   => \$read_length,
           'read_number=i'   => \$read_number,
           'error_rate=f'    => \$error_rate,
           'help!'           => \$help,
           );
pod2usage(-verbose => 1)  if ($help);
           pod2usage(-verbose => 2)  if ($man);
           unless ( defined( $genome_fasta ) && defined( $read_length ) )
                  {
                   Pod::Usage::pod2usage( -exitstatus => 1);
                  }

open (my $infile,"<",$genome_fasta)|| die "can't open file: $!";
local $/ = "\n>";  # read by FASTA record
my $Sequence;
my $basecount;
while (my$seq_chunk = <$infile>) {
       chomp $seq_chunk;
       $seq_chunk =~ s/^>*.+\n//;  # remove FASTA header
       $seq_chunk =~ s/\n//g;  # remove endlines
       $Sequence.=$seq_chunk;
      }
close($infile);
my $A=$Sequence=~tr/A/A/;
my $G=$Sequence=~tr/G/G/;
my $T=$Sequence=~tr/T/T/;
my $C=$Sequence=~tr/C/C/;
my $N=$Sequence=~tr/N/N/;
$basecount=($A+$G+$T+$C+$N)-$read_length; # -$read_length: to restrict max rand position and allow last bases to be simulated.

my $count=0;	
my @ranPOS = @{randomNUM($read_number,$basecount)};
open (my $outfile,">","pos.bed")||die "can't open $!";
open (my $outfile1,">","read.fq")||die "can't open $!";
open (my $outfile2,">","readSNV.fq")||die "can't open $!";	
      foreach my $pos(@ranPOS){
        print $outfile "1\t",$pos-1,"\t",$pos+$read_length-1,"\t","+","\n"; ## save read pos as 0-based bed file
	$count++;
        my $fasta=substr($Sequence,$pos,$read_length);
        print $outfile1 "\@HISEQ:".$count.":".$pos."#0/1","\n";
        print $outfile1 $fasta,"\n";
        print $outfile1 "+", "\n";
        print $outfile1 "?" x $read_length,"\n";
        my $snv = $read_number * $read_length * $error_rate; ## calculate no. of bases to be substituted in genome
        if ($count <= $snv){
            my $ran = int(rand($read_length-1));
            my $nuc = substr($fasta,$ran,1);
	    substr($fasta,$ran,1,randomSub($nuc));
            print $outfile2 "\@HISEQ:".$count.":".$pos."#0/1","\n";
            print $outfile2 $fasta,"\n";
            print $outfile2 "+", "\n";
            print $outfile2 "?" x $read_length,"\n";	
	    }
	else {
	      print $outfile2 "\@HISEQ:".$count.":".$pos."#0/1","\n";
              print $outfile2 $fasta,"\n";
              print $outfile2 "+", "\n";
              print $outfile2 "?" x $read_length,"\n";
	     }
	}
close ($outfile);
close ($outfile1);
close ($outfile2);
##Suroutine to return random positions in genome
################################################
sub randomNUM {
    my ($j,$k)= @_;
    my @arrRND;
    for (my $i=0;$i<$j;$i++){
         my $ranNUM = int(rand($k));	
         push @arrRND,$ranNUM;
    }
         return \@arrRND;
}

##Subroutine to return random substitue base: eg. if 'A' then return T or G or C
################################################################################
sub randomSub {
     my $base=$_[0];
     my @chars=("A","G","C","T");
        foreach my$b(@chars){
                if ($b eq "$base"){
                @chars=grep{!/^$base/}@chars;
                my $substitute = @chars[map{rand @chars}(1)];
                return $substitute;
                }
        }
}

=head1 SYNOPSIS

perl sim.pl [options]

  Options:

   --genome_fasta
   --read_length
   --read_number
   --error_rate
   
   Example:
        perl sim.pl --genome_fasta genome.fa --read_length 50 --read_number 100000 --error_rate 0.01

=head1 DESCRIPTION

##
This program simulates single end reads in fastq format from a given genome and also introduces random mutations. 

##

=head1 OPTIONS

=over 4


=item B<--genome_fasta>

Provide reference genome fasta file

=item B<--read_length>

Provide expected read length for simulated reads

=item B<--read_number>

Provide expected number of simulated reads to be produced

=item B<--error_rate>

Provide rate of mutation to be introduced in simulated reads

=back

=head1 COPYRIGHT

This program is free software and can redistributed and/or modified.

=cut

